File: src/test/kotlin/com/scully/CustomTestListener.java
package com.scully;

import com.codeborne.selenide.Screenshots;
import com.codeborne.selenide.WebDriverRunner;
import com.google.common.io.Files;
import io.qameta.allure.Attachment;
import org.testng.ITestContext;
import org.testng.ITestListener;
import org.testng.ITestResult;

import java.io.IOException;

public class CustomTestListener implements ITestListener {

  @Override
  public void onTestStart(ITestResult result) {

  }

  @Override
  public void onTestSuccess(ITestResult result) {
    try {
      this.screenShot();
    } catch (IOException e) {
      throw new RuntimeException("Error creating screenshot");
    }
    this.capturePageSource();
  }

  @Override
  public void onTestFailure(ITestResult result) {
    try {
      this.screenShot();
    } catch (IOException e) {
      throw new RuntimeException("Error creating screenshot");
    }
    this.capturePageSource();
  }

  @Override
  public void onTestSkipped(ITestResult result) {

  }

  @Override
  public void onTestFailedButWithinSuccessPercentage(ITestResult result) {

  }

  @Override
  public void onStart(ITestContext context) {

  }

  @Override
  public void onFinish(ITestContext context) {
  }


  @Attachment(type = "text/html")
  private String capturePageSource() {
    return WebDriverRunner.source();
  }

  @Attachment(type = "image/png")
  private byte[] screenShot() throws IOException {
    var screenShot = Screenshots.takeScreenShotAsFile();
    return Files.toByteArray(screenShot);
  }

}

File: src/test/kotlin/com/scully/listeners/CustomTestListener.kt
package com.scully.listeners

import com.codeborne.selenide.Screenshots
import com.codeborne.selenide.WebDriverRunner
import io.qameta.allure.Attachment
import org.testng.ITestContext
import org.testng.ITestListener
import org.testng.ITestResult
import java.nio.file.Files

class CustomTestListener : ITestListener {

    override fun onTestStart(result: ITestResult) {}

    override fun onTestSuccess(result: ITestResult) {
        `Screen shot`()
        `HTML Source Code`()
    }


    override fun onTestFailure(result: ITestResult) {
        `Screen shot`()
        `HTML Source Code`()
    }

    override fun onTestSkipped(result: ITestResult) {}

    override fun onTestFailedButWithinSuccessPercentage(result: ITestResult) {}

    override fun onStart(context: ITestContext) {}

    override fun onFinish(context: ITestContext) {}

    @Attachment(type = "text/html")
    private fun `HTML Source Code`(): String {
        return WebDriverRunner.source()
    }

    @Attachment(type = "image/png")
    private fun `Screen shot`(): ByteArray {
        val screenShot = Screenshots.takeScreenShotAsFile()
        return Files.readAllBytes(screenShot.toPath())
    }

}